mypy/checkexpr.py

```diff
@@ -1119,13 +1116,25 @@ def check_boolean_op(self, e: OpExpr, context: Context) -> Type:
         self.check_usable_type(left_type, context)
         self.check_usable_type(right_type, context)
 
-        # If either of the type maps is None that means that result cannot happen.
-        # If both of the type maps are None we just have no information.
-        if left_map is not None and right_map is None:
+        if right_map is None:
+            # The boolean expression is statically known to be the left value
+            assert left_map is not None  # find_isinstance_check guarantees this
             return left_type
-        elif left_map is None and right_map is not None:
+        if left_map is None:
+            # The boolean expression is statically known to be the right value
+            assert right_map is not None  # find_isinstance_check guarantees this
             return right_type
-        return UnionType.make_simplified_union([left_type, right_type])
+
+        restricted_left_type = truthiness_restriction(left_type)
+
+        if isinstance(restricted_left_type, UninhabitedType):
+            # The left operand can never be the result
+            return right_type
+        elif left_type is restricted_left_type:
```

Comment:
I think it would be better for the right hand expression to be analyzed regardless.  Right now there are some places where we don't analyze unreachable code, but that's a bug rather than a feature.